File: Queue/Double-Ended Queue/LinkedList-Based/DEQueueList.py
class Node:

    def __init__(self):
        self.data = None
        self.Next = None

class DEQueueList:

    def __init__(self, num):
        self.head = None
        self.size = 0
        self.MAX = num
        self.FRONT = -1
        self.REAR = -1

    #Check whether the Queue is Empty or not
    def isEmpty(self):
        return self.head == None

    #Check whether the Queue is Full or not
    def isFull(self):
        return self.size == self.MAX

    #Displaying The Queue
    def display(self):

        if self.isEmpty():
            print("Queue is Empty!");
            return

        currentNode = self.head

        for i in range(self.size):
            print(currentNode.data, end = " -> ")
            currentNode = currentNode.Next

        print()


    #Inserting Node(s)
    def insert(self, x, loc):

        newNode = Node()
        newNode.data = x

        if loc == 0:

            if self.isEmpty():

                self.head = newNode
                newNode.Next = self.head

            else:

                newNode.Next = self.head
                self.head = newNode

                currentNode = self.head

                for i in range(self.size):
                    currentNode = currentNode.Next

                currentNode.Next = self.head

        else:

            if loc > self.size:

                print("The Queue is only ", size, " nodes")
                return

            currentNode = self.head

            for i in range(loc - 1):
                currentNode = currentNode.Next

            newNode.Next = currentNode.Next
            currentNode.Next = newNode

        self.size += 1


    #Deleting Node(s)
    def delete(self, loc):

        if self.isEmpty():

            print("Queue is Empty!")
            return

        if loc > self.size - 1:

            print("The Queue is only ", self.size, "nodes")
            return

        currentNode = self.head
        previousNode = self.head

        if loc == 0:

            if self.size == 1:

                self.head = None

            else:

                self.head = currentNode.Next

                for i in range(self.size - 1):
                    currentNode = currentNode.Next

                currentNode.Next = self.head

        elif loc == self.size - 1:

            for i in range(loc - 1):
                currentNode = currentNode.Next

            currentNode.Next = self.head

        else:

            for i in range(loc):

                currentNode = currentNode.Next

                if i == 0:
                    continue

                previousNode = previousNode.Next

            previousNode.Next = currentNode.Next

        self.size -= 1

    #Insert At REAR
    def insertREAR(self, x):

        if self.isFull():
            print("Queue is Full!")
            return

        print("\nBefore Inserting At REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isEmpty():

            self.FRONT += 1
            self.REAR += 1

            self.insert(x, self.REAR)

        elif self.REAR == self.MAX - 1:

            self.REAR = 0

            self.insert(x, self.REAR)

        else:

            self.REAR += 1

            self.insert(x, self.REAR)

        print("After Inserting At REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Delete From REAR
    def deleteREAR(self):

        if self.isEmpty():
            print("Queue is Empty!")
            return

        print("\nBefore Deleting From REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        self.delete(self.REAR)

        if self.FRONT == self.REAR:

            self.FRONT = -1
            self.REAR = -1

        elif self.REAR == 0:

            self.REAR = self.size - 1

        else:

            self.REAR -= 1

        print("After Deleting From REAR: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

     #Insert At FRONT
    def insertFRONT(self, x):

        if self.isFull():
            print("Queue is Full!")
            return

        print("\nBefore Inserting At FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.isEmpty():

            self.FRONT += 1
            self.REAR += 1

            self.insert(x, self.FRONT)

        elif self.FRONT == 0:

            self.FRONT = self.size

            self.insert(x, self.FRONT)

        else:

            self.insert(x, self.FRONT + 1)

        print("After Inserting At FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


    #Delete From FRONT
    def deleteFRONT(self):

        if self.isEmpty():
            print("Queue is Empty!")
            return

        print("\nBefore Deleting From FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))

        if self.FRONT == self.REAR:

            self.delete(self.FRONT)

            self.FRONT = -1
            self.REAR = -1

        elif self.FRONT == self.size - 1:

            self.delete(self.FRONT)
            self.FRONT = 0

        elif self.FRONT > self.REAR and self.FRONT < self.size:

            self.delete(self.FRONT)

        else:

            self.delete(self.FRONT)
            self.REAR -= 1

        print("After Deleting From FRONT: FRONT = {} and REAR = {}".format(self.FRONT, self.REAR))


D = DEQueueList(5)

D.display()

D.display();

D.insertREAR(1);
D.display();

D.insertREAR(2);
D.display();

D.insertREAR(3);
D.display();

D.insertREAR(4);
D.display();

D.insertREAR(5);
D.display();

#insertREAR(q, 6); //OverFlow

D.deleteREAR();
D.display();

D.deleteREAR();
D.display();

D.insertFRONT(4);
D.display();

D.insertFRONT(5);
D.display();

#insertFRONT(q, 6); //OverFlow

D.deleteFRONT();
D.display();

D.deleteFRONT();
D.display();

D.deleteFRONT();
D.display();

D.deleteFRONT();
D.display();

D.deleteFRONT();
D.display();

File: Queue/Double-Ended Queue/LinkedList-Based/DEQueueList.java
class DEQueueList{

    //Establishing Node
    class Node {
        int data;
        Node next;
    }

    private Node head;
    private int size;
    private int MAX;
    private int FRONT;
    private int REAR;

    //Initialization
    DEQueueList(int num){

        head = null;
        size = 0;
        MAX = 5;
        FRONT = -1;
        REAR = -1;

    }

    //Check whether the Queue is Empty or not
    public boolean isEmpty(){
        return head == null;
    }

    //Check whether the Queue is Full or not
    public boolean isFull(){
        return size == MAX;
    }


    //Displaying The Queue
    public void display(){

        if(isEmpty()){
            System.out.println("Queue is Empty!\n");
            return;
        }

        Node currentNode = head;

        do {

            System.out.print(currentNode.data + " -> ");
            currentNode = currentNode.next;

        } while(currentNode != head);

        System.out.println();

    }

    //Inserting Node(s)
    void insert(int x, int loc){

        Node newNode = new Node();
        newNode.data = x;

        if(loc == 0){

            if(isEmpty()){

                head = newNode;
                newNode.next = head;

            } else {

                newNode.next = head;
                head = newNode;

                Node currentNode = head;

                for(int i = 0; i < size; i++)
                    currentNode = currentNode.next;

                currentNode.next = head;

            }

        } else {

            if(loc > size){

                System.out.println("The Queue is only " + size + " nodes!");
                return;

            }

            Node currentNode = head;

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            newNode.next = currentNode.next;
            currentNode.next = newNode;

        }

        size++;

    }

    //Deleting Node(s)
    void delete(int loc){

        if(isEmpty()){

            System.out.println("Queue is Empty!\n");
            return;

        }

        if(loc > size - 1){

            System.out.println("The Queue is only " + size + " nodes!");
            return;

        }

        Node currentNode = head;
        Node previousNode = head;

        if(loc == 0){

            if(size == 1)
                head = null;
            else {

                head = currentNode.next;

                for(int i = 0; i < size - 1; i++)
                    currentNode = currentNode.next;

                currentNode.next = head;

            }

        } else if(loc == size - 1){

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            currentNode.next = head;

        } else {

            for(int i = 0; i < loc; i++){

                currentNode = currentNode.next;

                if(i == 0)
                    continue;

                previousNode = previousNode.next;

            }

            previousNode.next = currentNode.next;

        }

        size--;

    }

    public void insertREAR(int x){

        if(isFull()){
            System.out.println("Queue is Full!");
            return;
        }

        System.out.println("\nBefore Inserting At REAR: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isEmpty()){
            FRONT++;
            REAR++;

            insert(x, REAR);

        } else if(REAR == MAX - 1){

            REAR = 0;
            insert(x, REAR);

        } else {

            REAR++;
            insert(x, REAR);

        }

        System.out.println("\nAfter Inserting At REAR: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Deleting From REAR
    public void deleteREAR(){

        if(isEmpty()){
            System.out.println("Queue is Full!");
            return;
        }

        System.out.println("\nBefore Deleting From REAR: FRONT = " + FRONT + " and REAR = " + REAR);
        delete(REAR);

        if(FRONT == REAR){

            FRONT = -1;
            REAR = -1;

        } else if(REAR == 0)
            REAR = size - 1;
        else
            REAR--;

        System.out.println("After Deleting From REAR: FRONT = " + FRONT + " and REAR = " + REAR);


    }

    //Inserting At FRONT
    public void insertFRONT(int x){

        if(isFull()){
            System.out.println("Queue is Full!");
            return;
        }

        System.out.println("\nBefore Inserting At FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

        if(isEmpty()){

            FRONT++;
            REAR++;

            insert(x, FRONT);

        } else if(FRONT == 0){

            FRONT = size;

            insert(x, FRONT);

        } else {

            int temp = FRONT;

            insert(x, ++temp);

        }

        System.out.println("\nAfter Inserting At FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

    }

    //Deleting From FRONT
    public void deleteFRONT(){

        if(isEmpty()){
            System.out.println("Queue is Empty!");
            return;
        }

        System.out.println("\nBefore Deleting From FRONT: FRONT = " + FRONT + " and REAR = " + REAR);
        if(FRONT == REAR){

            delete(FRONT);

            FRONT = -1;
            REAR = -1;

        } else if(FRONT == size - 1){

            delete(FRONT);
            FRONT = 0;

        } else if(FRONT > REAR && FRONT < size){

            delete(FRONT);

        } else {

            delete(FRONT);
            REAR--;

        }

        System.out.println("After Deleting From FRONT: FRONT = " + FRONT + " and REAR = " + REAR);

    }

}
